File: leetcode/AIcoding/1.Gemini-2.5/cpp_solutions/930-binary-subarrays-with-sum.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    int numSubarraysWithSum(std::vector<int>& nums, int goal) {
        int count = 0;
        int current_sum = 0;
        std::unordered_map<int, int> freq;
        freq[0] = 1;

        for (int num : nums) {
            current_sum += num;
            if (freq.count(current_sum - goal)) {
                count += freq[current_sum - goal];
            }
            freq[current_sum]++;
        }

        return count;
    }
};

File: leetcode/AIcoding/1.Gemini-2.5/rust_solutions/930-binary-subarrays-with-sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn num_subarrays_with_sum(nums: Vec<i32>, goal: i32) -> i32 {
        let mut count = 0;
        let mut current_sum = 0;
        let mut prefix_sums: HashMap<i32, i32> = HashMap::new();

        prefix_sums.insert(0, 1);

        for num in nums {
            current_sum += num;

            if let Some(&freq) = prefix_sums.get(&(current_sum - goal)) {
                count += freq;
            }

            *prefix_sums.entry(current_sum).or_insert(0) += 1;
        }

        count
    }
}